crates/bevy_asset/src/path.rs

```diff
@@ -63,6 +63,66 @@ impl<'a> AssetPath<'a> {
                 .map(|value| Cow::Owned(value.to_string())),
         }
     }
+
+    /// Resolves a possibly-relative path string relative to a base path, producing an `AssetPath`.
+    ///
+    /// If the path argument begins with `#`, then it is considered an asset label, in which case
+    /// the result is the base path with the label portion replaced.
+    ///
+    /// If the path argument starts with "./" or "../", then the result the concatenation of
+    /// the base path and the path argument, which is then canonicalized. The rules are the same as
+    /// for URIs, which means that the path argument is resolved relative to the *directory* of the
+    /// base path, so a base path of `"x/y/z#foo"` combined with a relative path of `"./a#bar"`
+    /// yields `"x/y/a#bar"`.
+    ///
+    /// If neither of the above are true, then the path argument is considered a 'full' path,
+    /// and the result is simply a canonicalized version of the argument string, converted
+    /// into an `AssetPath`. Note that a 'full' asset path is still relative to the asset root
+    /// directory, and not necessarily an absolute filesystem path.
+    ///
+    /// Note that "./" and "../" elements are only canonicalized in the path argument, any such
+    /// elements in the base path are left as-is.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the path argument has more 'parent' elements ("../") than
+    /// are available in the base path.
+    pub fn resolve(&'a self, path: &'a str) -> AssetPath<'a> {
+        if let Some(label) = path.strip_prefix('#') {
+            // It's a label only
+            AssetPath::new_ref(&self.path, Some(label))
+        } else {
+            let (rpath, rlabel) = match path.split_once('#') {
+                Some((path, label)) => (path, Some(label.to_string())),
+                None => (path, None),
+            };
+            let mut fpath = PathBuf::from(self.path());
+            if !fpath.pop() {
+                panic!("Can't compute relative path - not enough path elements");
+            }
+
+            let rpath = PathBuf::from(rpath);
+            let mut first = true;
+            for elt in rpath.iter() {
```

Comment:
Nit: I think using [`components`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.components) might be a bit more readable and robust to any potential OS differences. This way you can specifically reference `Component::CurDir` and `Component::ParentDir`.

Reply:
I decided not to do this because I wasn't sure what the knock-on effects would be.

Because asset paths are used both as in-memory keys as well as filesystem paths, it's important that platform idiosyncrasies like drive letters and path separators don't leak into the canonical representation of the asset paths. Because I'm not that familiar with the APIs you've proposed, I wasn't sure what the effects would be.